n2survey/lime/survey.py

```diff
@@ -280,6 +306,115 @@ def get_responses(
+            )
+
+    def keep_choices(
+        self, question: str, choices: Union[str, bool, list], logic: str = "or"
+    ) -> "LimeSurvey":
+        """Filter responses DataFrame by keeping certain choices to specified questions
+
+        Args:
+            conditions (tuple or list of tuples): Conditions to re-format. Tuples in
+                format of (question_id, choices). choices can be a single str or list
+                of str. Multiple conditions are input as a list of tuples.
+                E.g. ("A6", "Woman") or [("A6", ["Woman", "Man"]), ("B6_SQ001", "Yes")]
+
+        Returns:
+            LimeSurvey: LimeSurvey with filtered responses
+        """
+
+        logic_dict = {"and": " & ", "or": " | "}
+        question_type = self.get_question_type(question)
+
+        # Convert into list
+        if not isinstance(choices, list):
+            choices = [choices]
+
+        # Handle multiple-choice questions with choices as sub-question
+        # labels differently, e.g. keep_choices("A3", "Chemistry")
+        if (question_type == "multiple-choice") & (True not in choices):
+            inverted_answer_dict = {
+                answer: id for id, answer in self.get_choices(question).items()
+            }
+            subquestions = [inverted_answer_dict[choice] for choice in choices]
+            return self.query(
+                logic_dict[logic].join(
+                    [f"{subquestion} == 'Y'" for subquestion in subquestions]
+                )
+            )
+        # All other cases, including single-choice questions, array sub-questions,
+        # mutliple-choice sub-questions with True labels
+        else:
+            return self.filter(
+                self.get_responses(question, labels=False)
+                .isin(choices)
+                .any(axis="columns")
+                | self.get_responses(question, labels=True)
+                .isin(choices)
+                .any(axis="columns")
+            )
+
+    def filter(self, mask: Union[pd.Series, pd.DataFrame]) -> "LimeSurvey":
+        """Filter responses DataFrame by applying a mask Series or DataFrame
+
+        Args:
+            mask (pd.Series): A bool Series or DataFrame as mask
+
+        Returns:
+            LimeSurvey: LimeSurvey with filtered responses
+        """
+
+        # Make copy of LimeSurvey instance
+        filtered_survey = self.__copy__()
+        # Filter responses DataFrame
+        filtered_survey.responses = self.responses[mask]
+
+        return filtered_survey
+
+    def query(self, expr: str) -> "LimeSurvey":
+        """Filter responses DataFrame with a boolean expression
+
+        Args:
+            expr (str): Condition str for pd.DataFrame.query().
+                E.g. "A6 == 'Woman' & "C3 == 'I do not like my topic.'"
+
+        Returns:
+            LimeSurvey: LimeSurvey with filtered responses
+        """
+
+        # Make copy of LimeSurvey instance
+        filtered_survey = self.__copy__()
```

Comment:
Please, try `self.copy()`

Reply:
This doesn't work. I get "'LimeSurvey' object has no attribute 'copy'" error.